test/integration/master/audit_dynamic_test.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	auditregv1alpha1 "k8s.io/api/auditregistration/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	auditv1 "k8s.io/apiserver/pkg/apis/audit/v1"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/client-go/kubernetes"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+	"k8s.io/kubernetes/test/utils"
+)
+
+// TestDynamicAudit ensures that v1alpha of the auditregistration api works
+func TestDynamicAudit(t *testing.T) {
+	// start api server
+	result := kubeapiservertesting.StartTestServerOrDie(t, nil,
+		[]string{
+			"--audit-dynamic-configuration",
+			"--feature-gates=DynamicAuditing=true",
+		},
+		framework.SharedEtcd())
+	defer result.TearDownFn()
+
+	// create blank atomic event lists
+	eventList1 := &atomic.Value{}
+	eventList1.Store(auditinternal.EventList{})
+	eventList2 := &atomic.Value{}
+	eventList2.Store(auditinternal.EventList{})
+
+	// start test servers
+	server1 := httptest.NewServer(buildTestHandler(t, eventList1))
+	defer server1.Close()
+	server2 := httptest.NewServer(buildTestHandler(t, eventList2))
+	defer server2.Close()
+
+	// check that servers are healthy
+	err := testServerHealthCheck(server1.URL)
+	require.NoError(t, err)
+	err = testServerHealthCheck(server2.URL)
+	require.NoError(t, err)
+
+	kubeclient, err := kubernetes.NewForConfig(result.ClientConfig)
+	require.NoError(t, err)
+
+	// build audit sinks
+	sink1 := buildAuditSink("test1", server1.URL)
+	sink2 := buildAuditSink("test2", server2.URL)
+
+	t.Run("one sink", func(t *testing.T) {
+		_, err := kubeclient.AuditregistrationV1alpha1().AuditSinks().Create(&sink1)
+		require.NoError(t, err, "failed to create audit sink1")
+		t.Log("created audit sink1")
+
+		// perform configmap ops
+		configMapOperations(t, kubeclient)
```

Comment:
performing these immediately after registering the sink races with the dynamic backend observing the sink and registering the handler, right? need to do one of the following:
* wait for positive confirmation the sink is active (which we can't do because we don't expose that in the API)
* wait long enough after registering the sink that we would consider not observing it in that period a failure (5 seconds? 10 seconds?)
* retry the "configMapOperations && checkForEvents" operation in a loop until it passes or we exceed the time we consider acceptable for the sink to have become active

same comment applies other places a sink is added/removed/updated

Reply:
This is similar to my comment about a potential flake risk. I think we could start by waiting for 5 seconds after the sink has been added/updated/deleted before performing operations. I talked with @pbarker about the possibility of doing this in a loop, but he wasn't sure if we could get to a clean state after running the loop multiple times. We can continue to discuss if needed.

Would you be ok with a sleep for starters?

Reply:
> Would you be ok with a sleep for starters?

I could live with a 5 second timeout as long as you watch to make sure we're not hitting flakes, and look into a follow up that makes it more deterministic.

That said, I just looked at how long this new test is taking. These are the five run times for the test/integration/master package before this PR:
```
335.342s
323.410s
335.549s
336.179s
359.685s
```

and with this PR:
```
464.194s
449.109s
450.238s
467.853s
468.212s
```

This new test seems to be adding ~2 minutes to the test/integration/master package run time. That seems excessive (and long run times can make tests flake on timeouts as well). Is that runtime expected? Can we get it closer to ~20 seconds? The other 25 tests in the package complete much more quickly

Reply:
ok this is updated, this was caused by waiting on the buffers to flush because we hadn't hit the max batch or wait time. Implemented @liggitt's req and moved the example test server into the framework, and modified the audit options so that the batch config could be hacked. Test takes about 13s now :)